### PR TITLE
Remove lang restriction

### PR DESCRIPTION
**Pull request by nuss**
This pull request adds a multi-language version of the plugin, including German and English. Other languages can be added by adding appropriate .mo/.po files.
Within the settings page one can now select between available translation languages (including translated example texts).
In the database the settings are stored in separate entries for each language, denoted by their locale. E.g.:
`akismet_privacy_notice_settings_en_US` (American English)
`akismet_privacy_notice_settings_de_DE` (German)
On a multilingual page the privacy notice will be displayed according to the currently selected language.

### CODE REVIEW

File: akismet-privacy-policies.php [hunk 1]
```diff
@@ -38,33 +46,49 @@ class Akismet_Privacy_Policies {
...
+
+		if ( isset( $_GET[ 'translation' ])) {
+			$this->translation = $_GET[ 'translation' ];
+		} elseif( isset( $_POST[ 'translation' ])) {
+			$this->translation = $_POST[ 'translation' ];
+		} else {
+			$this->translation = get_user_locale();
+		}
+
+		if ( $this->translation != 'en_US' ) {
```
Comment (bueltge) [inline akismet-privacy-policies.php]: Please check always strict for clear coding, speed --> `!==`. However in this context, it is also a idea to leave `en_US` for other language keys wit `en`? So like `en_ NZ` have also the chance to use this translation content?

File: akismet-privacy-policies.php [hunk 2]
```diff
@@ -38,33 +46,49 @@ class Akismet_Privacy_Policies {
...
+			$this->mo = new Mo;
+			$mofile = dirname( __FILE__) . '/languages/akismet-privacy-policies-' . $this->translation . '.mo';
```
Comment (bueltge) [inline akismet-privacy-policies.php]: Need a space after `__FILE__`.

File: akismet-privacy-policies.php [hunk 3]
```diff
@@ -86,6 +110,30 @@ public static function get_object() {
...
+	 * Initialize and translate $this->notice, $this->error_message
+	 *
+	 * @since x.x.x
+	 * @access public
+	 *
+	 * @return void
+	 */
+	public function translate_strings() {
+		// dummy translation, only there to let poedit recognize the text...
+		__( 'I accept that my given data and my IP address is sent to a server in the USA only for the purpose of spam prevention through the <a href="http://akismet.com/">Akismet</a> program.<br /><a href="https://akismet.com/gdpr/">More information on Akismet and GDPR</a>.', 'akismet-privacy-policies' );
```
Comment (bueltge) [inline akismet-privacy-policies.php]: We need escaping so we haven' control about the content inside the language files, so please switch for the example from `__()` to `esc_html__()`.

File: akismet-privacy-policies.php [hunk 4]
```diff
@@ -86,6 +110,30 @@ public static function get_object() {
...
+    __( '<p><strong>Attention:</strong> You have not accepted our privacy disclaimer.</p>', 'akismet-privacy-policies' );
+		if ( $this->translation != 'en_US' ) {
+			$this->notice = $this->mo->translate( 'I accept that my given data and my IP address is sent to a server in the USA only for the purpose of spam prevention through the <a href="http://akismet.com/">Akismet</a> program.<br /><a href="https://akismet.com/gdpr/">More information on Akismet and GDPR</a>.', 'akismet-privacy-policies' );
+		} else {
```
Comment (bueltge) [inline akismet-privacy-policies.php]: The else control is not necessary, so we can remove them.

File: akismet-privacy-policies.php [hunk 5]
```diff
@@ -86,6 +110,30 @@ public static function get_object() {
...
+			$this->notice = 'I accept that my given data and my IP address is sent to a server in the USA only for the purpose of spam prevention through the <a href="http://akismet.com/">Akismet</a> program.<br /><a href="https://akismet.com/gdpr/">More information on Akismet and GDPR</a>.';
+		}
+		if ( $this->translation != 'en_US' ) {
+			$this->error_message = $this->mo->translate( '<p><strong>Attention:</strong> You have not accepted our privacy disclaimer.</p>', 'akismet-privacy-policies' );
+		} else {
```
Comment (bueltge) [inline akismet-privacy-policies.php]: like above, the else is not necessary.

File: akismet-privacy-policies.php [hunk 6]
```diff
@@ -86,6 +110,30 @@ public static function get_object() {
...
+		__( 'I accept that my given data and my IP address is sent to a server in the USA only for the purpose of spam prevention through the <a href="http://akismet.com/">Akismet</a> program.<br /><a href="https://akismet.com/gdpr/">More information on Akismet and GDPR</a>.', 'akismet-privacy-policies' );
+    __( '<p><strong>Attention:</strong> You have not accepted our privacy disclaimer.</p>', 'akismet-privacy-policies' );
+		if ( $this->translation != 'en_US' ) {
+			$this->notice = $this->mo->translate( 'I accept that my given data and my IP address is sent to a server in the USA only for the purpose of spam prevention through the <a href="http://akismet.com/">Akismet</a> program.<br /><a href="https://akismet.com/gdpr/">More information on Akismet and GDPR</a>.', 'akismet-privacy-policies' );
+		} else {
+			$this->notice = 'I accept that my given data and my IP address is sent to a server in the USA only for the purpose of spam prevention through the <a href="http://akismet.com/">Akismet</a> program.<br /><a href="https://akismet.com/gdpr/">More information on Akismet and GDPR</a>.';
+		}
+		if ( $this->translation != 'en_US' ) {
```
Comment (bueltge) [inline akismet-privacy-policies.php]: strict please `!==`.

File: akismet-privacy-policies.php [hunk 7]
```diff
@@ -123,19 +184,21 @@ public function add_comment_notice( $arr_comment_defaults ) {
 			return $arr_comment_defaults;
 		}
 
-		$options = get_option( 'akismet_privacy_notice_settings' );
-		if ( ! isset( $options[ 'checkbox' ] ) || empty( $options[ 'checkbox' ] ) && 0 != $options[ 'checkbox' ] ) {
-			$options[ 'checkbox' ] = $this->checkbox;
+		// $locale = isset( $_GET[ 'translation' ]) ? $_GET[ 'lang'] : get_locale();
+		// $this->options = get_option( 'akismet_privacy_notice_settings_' . $this->translation );
+
+		if ( ! isset( $this->options[ 'checkbox' ] ) || empty( $this->options[ 'checkbox' ] ) && 0 != $this->options[ 'checkbox' ] ) {
```
Comment (bueltge) [inline akismet-privacy-policies.php]: strict please, `!==`

File: akismet-privacy-policies.php [hunk 8]
```diff
@@ -123,19 +184,21 @@ public function add_comment_notice( $arr_comment_defaults ) {
...
+			$this->options[ 'checkbox' ] = $this->checkbox;
 		}
-		if ( empty( $options[ 'notice' ] ) ) {
-			$options[ 'notice' ] = $this->notice;
+		if ( empty( $this->options[ 'notice' ] ) ) {
+			$this->options[ 'notice' ] = __( $this->notice );
```
Comment (bueltge) [inline akismet-privacy-policies.php]: Need escaping.

File: akismet-privacy-policies.php [hunk 9]
```diff
@@ -187,14 +250,13 @@ public function error_message() {
...
-		if ( empty( $options[ 'error_message' ] ) ) {
-			$options[ 'error_message' ] = $this->error_message;
+		if ( empty( $this->options[ 'error_message' ] ) ) {
+			$this->options[ 'error_message' ] = $this->error_message;
 		}
 
 		// check for checkbox active
 		if ( isset( $_POST[ 'comment' ] ) && ( ! isset( $_POST[ 'akismet_privacy_check' ] ) ) ) {
-			$message = apply_filters( 'akismet_privacy_error_message', $options[ 'error_message' ] );
+			$message = apply_filters( 'akismet_privacy_error_message', __( $this->options[ 'error_message' ], 'akismet-privacy-policies' ) );
```
Comment (bueltge) [inline akismet-privacy-policies.php]: Need escaping.

File: akismet-privacy-policies.php [hunk 10]
```diff
@@ -278,84 +339,112 @@ public function get_settings_page() {
...
-						<td>
-							<textarea id="akismet_privacy_error_message" name="akismet_privacy_notice_settings[error_message]" cols="80"
-								rows="10" aria-required="true"><?php if ( isset( $options[ 'error_message' ] ) ) {
-									echo $options[ 'error_message' ];
+								<br /><?php _e( '<strong>Note:</strong> HTML is possible', 'akismet-privacy-policies' ) ?>
+								<br /><?php _e( '<strong>Attention:</strong> You will have to add the link to your privacy statement manually. In Wordpress 5 and later you can find a link to a guide for creating your own privacy statement under \'Settings\' &rarr; \'Privacy\'.', "akismet-privacy-policies" ) ?>
+								<br /><strong><?php _e( 'Example:', 'akismet-privacy-policies' ) ?></strong> <?php
+								if ( $this->translation != 'en_US' ) {
+									$example_notice = $this->mo->translate( $this->notice );
+								} else {
```
Comment (bueltge) [inline akismet-privacy-policies.php]: The else control is not necessary.

File: akismet-privacy-policies.php [hunk 11]
```diff
@@ -278,84 +339,112 @@ public function get_settings_page() {
...
+									$example_notice = $this->notice;
+								}
+								echo esc_html( $example_notice );
```
Comment (bueltge) [inline akismet-privacy-policies.php]: Please use `esc_html_e()` to echo and sanitize the output.

**Comment by bueltge** [comment]
In this context of PR we should also add a new wiki entry for the language EN - https://github.com/inpsyde/Akismet-Privacy-Policies/wiki

**Comment by bueltge** [comment]
Also please add a note about this new feature, change inside the CHANGELOG file. Thx :heart: @nuss

**Comment by nuss** [comment]
> Hi @nuss
> thanks a lot for this work, your PR. Great work.
> However please read my small review. Especially please note about escaping for all strings, output of language files and check also that you always use strict control, like `!==` instead `!=`. Here and there is a else control not useful and reduce the code structure. I will see forward for your additional changes and get feedback again.

@bueltge Thanks very much for your review!

I have changed all occurrences of `!=` to `!==`. 

> Have you also a function inside the code to remove the entries inside the database on de-installation of de-activation of the plugin for the language strings?

Yes, that should be handled by `unregister_settings`: The function first collects the names of all options currently stored in the database into a variable `$all_options` by calling `wp_load_alloptions`. Then it selects those belonging to the plugin through a regular expression.
```php
$to_be_deleted = preg_grep( '/^akismet_privacy_notice_settings(_)*[a-z]*(_)*[A-Z]*$/', array_keys( $all_options ) );
```
Finally it iterates over the list of selected options, unregisters and deletes them.
```php
foreach( $to_be_deleted as $option ) {
  unregister_setting( 'akismet_privacy_notice_settings_group', $option );
  delete_option( $option );
}
```
I have tested installing and un/installing a couple of times and it worked for me as it should.

I am now going to check the other issues you have commented on, so, stay tuned. Thanks!

File: akismet-privacy-policies.php (hunk 1 above)
Comment (nuss) [reply akismet-privacy-policies.php]: How could this be handled efficiently? Are there special versions of WP for English speaking countries other than US? I have chosen `en_US` as I assumed this to be the default language and all other versions to be translations through `.po` files. Should the language test possibly be handled through a regular expression to filter out all possible versions of English?

File: akismet-privacy-policies.php (hunk 2 above)
Comment (nuss) [reply akismet-privacy-policies.php]: Done. There were a couple of other places where I forgot to put a space between openening/ending brackets and the text within the brackets. I will correct them all.

File: akismet-privacy-policies.php (hunk 3 above)
Comment (nuss) [reply akismet-privacy-policies.php]: Done

**Comment by bueltge** [comment]
Thanks, I see forward for your enhancement.

File: akismet-privacy-policies.php (hunk 4 above)
Comment (nuss) [reply akismet-privacy-policies.php]: I think this is a special case. If I leave out the `else` branch the English template strings are not displayed. The English template strings are not held in a `.po` file - they're defined in the php source. Thus, I'm assigning the untranslated version of the template string.
The `$this->mo->translate()` may seem a bit awkward. The idea behind is, that, if e.g. the admin interface's language has been switched to English, the template strings should still appear in the current translation language.

File: akismet-privacy-policies.php (hunk 5 above)
Comment (nuss) [reply akismet-privacy-policies.php]: See my previous comment.

File: akismet-privacy-policies.php (hunk 6 above)
Comment (nuss) [reply akismet-privacy-policies.php]: Done

File: akismet-privacy-policies.php (hunk 7 above)
Comment (nuss) [reply akismet-privacy-policies.php]: Done

File: akismet-privacy-policies.php (hunk 10 above)
Comment (nuss) [reply akismet-privacy-policies.php]: I get an `undefined variable` warning if I leave out the `else` branch. Assigning `$example_notice` without language condition results in a fatal error...

File: akismet-privacy-policies.php (hunk 11 above)
Comment (nuss) [reply akismet-privacy-policies.php]: Yes, better. Thanks.

File: akismet-privacy-policies.php (hunk 8 above)
Comment (nuss) [reply akismet-privacy-policies.php]: Done

> In this context of PR we should also add a new wiki entry for the language EN - https://github.com/inpsyde/Akismet-Privacy-Policies/wiki

I'll do that as soon as I find time.

File: akismet-privacy-policies.php (hunk 9 above)
Comment (nuss) [reply akismet-privacy-policies.php]: Done

> Also please add a note about this new feature, change inside the CHANGELOG file. Thx heart @nuss

Oh! I think you've added the changelog after I had forked the repo, so it's missing in my fork. I'll figuere it out... Thanks very much!

**Comment by bueltge** [comment]
Hi @nuss 
I have merge for now in a new branch, so that I can check via IDE and publish a new version.
I will send you a information after done that you help again for testing before we release on w.org.
Thanks a lot!

**Pull request merged**